src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+        curl_multi_cleanup(m_handle);
+
+    // Global CURL cleanup if this is the last session
+    --ms_activeSessions;
+    if ( ms_activeSessions == 0 )
+        CleanupCURL();
+}
+
+void wxWebSessionCURL::Initialize()
+{
+    m_handle = curl_multi_init();
+}
+
+wxWebRequest* wxWebSessionCURL::CreateRequest(const wxString& url, int id)
+{
+    if ( !m_handle )
+        Initialize();
+
+    return new wxWebRequestCURL(*this, id, url);
+}
+
+static CURLMcode wx_curl_multi_wait(CURLM *multi_handle, int timeout_ms,
+    int *ret)
+{
+    // since libcurl 7.28.0 the curl_multi_wait method is more convient than
+    // calling multiple curl_multi_... methods.
+    // When support for older libcurl versions is dropped this wrapper can be
+    // removed.
+#if wxCURL_HAVE_MULTI_WAIT
+    return curl_multi_wait(multi_handle, NULL, 0, timeout_ms, ret);
+#else
+    wxASSERT(ret != NULL);
+
+    fd_set fdread;
+    fd_set fdwrite;
+    fd_set fdexcep;
+    timeval timeout;
+
+    long curl_timeo;
+
+    curl_multi_timeout(multi_handle, &curl_timeo);
+    if ( curl_timeo < 0 )
+        curl_timeo = timeout_ms;
+
+    timeout.tv_sec = curl_timeo / 1000;
+    timeout.tv_usec = (curl_timeo % 1000) * 1000;
+
+    FD_ZERO(&fdread);
+    FD_ZERO(&fdwrite);
+    FD_ZERO(&fdexcep);
+
+    curl_multi_fdset(multi_handle, &fdread, &fdwrite, &fdexcep, ret);
+    if ( *ret == -1 )
+        return CURLM_OK;
+    else if ( select(*ret + 1, &fdread, &fdwrite, &fdexcep, &timeout) == -1 )
+        return CURLM_BAD_SOCKET;
+    else
+        return CURLM_OK;
+#endif
+}
+
+wxThread::ExitCode wxWebSessionCURL::Entry()
+{
+    m_mutex.Lock();
+
+    int activeRequests = -1;
+    int repeats = 0;
+
+    while ( activeRequests )
+    {
+        // Handle cancelled requests
+        {
+            wxMutexLocker lock(m_cancelledMutex);
+            while ( !m_cancelledRequests.empty() )
+            {
+                wxObjectDataPtr<wxWebRequestCURL> request(m_cancelledRequests.back());
+                m_cancelledRequests.pop_back();
+                curl_multi_remove_handle(m_handle, request->GetHandle());
+                request->SetState(wxWebRequest::State_Cancelled);
+            }
+        }
+
+        // Instruct CURL to work on requests
+        curl_multi_perform(m_handle, &activeRequests);
+
+        // Process CURL message queue
+        int msgQueueCount;
+        while ( CURLMsg* msg = curl_multi_info_read(m_handle, &msgQueueCount) )
+        {
+            if ( msg->msg == CURLMSG_DONE )
+            {
+                wxWebRequestCURL* request;
+                curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &request);
+                curl_multi_remove_handle(m_handle, msg->easy_handle);
+                request->HandleCompletion();
+            }
+        }
+
+        if ( activeRequests )
+        {
+            // Wait for CURL work to finish
+            int numfds;
+            wx_curl_multi_wait(m_handle, 500, &numfds);
+
+            if ( !numfds )
```

Comment:
Could you please add a comment explaining why does this code do what it does? I kind of see what it is, but really not why...